README.rst
django-template-analyzer
========================

.. image:: https://github.com/edoburu/django-template-analyzer/actions/workflows/tests.yaml/badge.svg?branch=master
    :target: https://github.com/edoburu/django-template-analyzer/actions/workflows/tests.yaml
.. image:: https://img.shields.io/pypi/v/django-template-analyzer.svg
    :target: https://pypi.python.org/pypi/django-template-analyzer/
.. image:: https://img.shields.io/badge/wheel-yes-green.svg
    :target: https://pypi.python.org/pypi/django-template-analyzer/
.. image:: https://img.shields.io/codecov/c/github/edoburu/django-template-analyzer/master.svg
    :target: https://codecov.io/github/edoburu/django-template-analyzer?branch=master

The ``template_analyzer`` package offers an API to analyze the Django template structure.
It can be used to find nodes of a particular type, e.g. to do automatic detection of placeholder tags.

Supported features
==================

The scanner finds tags in various situations, including:

* Extend nodes
* Include nodes
* Overwritten blocks with new definitions
* Blocks with or without ``{{ block.super }}``
* Reorganized blocks
* Ignoring nodes outside blocks in extending templates
* Handling multiple levels of super includes

The returned nodes are provided in a natural ordering,
as they would be expected to appear in the outputted page.

While Django offers a ``template.nodelist.get_nodes_of_type()`` function,
this function does not produce the same results.


API example
===========

.. code-block:: python

    from django.template.loader import get_template
    from mycms.templatetags.placeholdertags import Placeholder
    from template_analyzer.djangoanalyzer import get_node_instances

    # Load a Django template
    template = get_template("mycms/default-page.html")

    # Find all tags in the template:
    placeholders = get_node_instances(template, Placeholder)

    # Read information from the template tag themselves:
    # (this is an example, accessing a custom method on the Placeholder object)
    placeholder_names = [p.get_name() for p in placeholders]

Installation
============

First install the module, preferably in a virtual environment. It can be installed from PyPI::

    pip install django-template-analyzer

Or the current folder can be installed::

    pip install .

Credits
=======

* This package is based on the work of
  `Django CMS <http://www.django-cms.org>`_. 
* Many thanks to the contributors of ``cms/utils/placeholder.py`` / ``cms/utils/plugins.py`` in Django CMS!


CHANGES.rst
Changelog
=========

Version 2.1 (2023-10-16)
------------------------

* Added Django 4.1/4.2 support

Version 2.0 (2021-11-16)
------------------------

* Dropped Python 2.7 support.
* Dropped Django 1.3-1.7 support.
* Confirmed Django 4.0 support.
* Reformatted the code with black and isort
* Migated to GitHub actions

Version 1.6.2 (2020-01-04)
--------------------------

* Fixed Django 3.0 compatibility by removing ``django.utils.six`` dependency.

Version 1.6.1
-------------

* Fixed breaking templates when using the cached template loader; content was shown multiple times.
* fixed support for parsing nodes from a custom template engine.
* Improve error messages for invalid template block names.

Version 1.6
-----------

* Added Django 1.9 support

Version 1.5
-----------

* Added Django 1.8 support

Version 1.4
-----------

* Added Django 1.7 support

Version 1.3
-----------

* Added Python 3 support

Version 1.2
-----------

* Support variable extends with defaults in templates, e.g. ``{% extends FOO|default:'base.html' %}``.

Version 1.1
-----------

* Fix Django 1.4 support

Version 1.0
-----------

* First release, based on the code of django-cms.


template_analyzer/tests.py
from django.template.base import TemplateSyntaxError
from django.template.loader import get_template
from django.test.testcases import TestCase


def get_placeholders(filename):
    from template_analyzer.djangoanalyzer import get_node_instances
    from template_analyzer.templatetags.template_analyzer_test_tags import Placeholder

    template = get_template(filename)

    # Definitely not the same:
    #placeholders = template.nodelist.get_nodes_by_type(Placeholder)

    placeholders = get_node_instances(template, Placeholder)
    return [p.get_name() for p in placeholders]


class PlaceholderTestCase(TestCase):

    def test_placeholder_scanning_extend(self):
        placeholders = get_placeholders('placeholder_tests/test_one.html')
        self.assertEqual(sorted(placeholders), sorted([u'new_one', u'two', u'three']))

    def test_placeholder_scanning_include(self):
        placeholders = get_placeholders('placeholder_tests/test_two.html')
        self.assertEqual(sorted(placeholders), sorted([u'child', u'three']))

    def test_placeholder_scanning_double_extend(self):
        placeholders = get_placeholders('placeholder_tests/test_three.html')
        self.assertEqual(sorted(placeholders), sorted([u'new_one', u'two', u'new_three']))

    def test_placeholder_scanning_complex(self):
        placeholders = get_placeholders('placeholder_tests/test_four.html')
        self.assertEqual(sorted(placeholders), sorted([u'new_one', u'child', u'four']))

    def test_placeholder_scanning_super(self):
        placeholders = get_placeholders('placeholder_tests/test_five.html')
        self.assertEqual(sorted(placeholders), sorted([u'one', u'extra_one', u'two', u'three']))

    def test_placeholder_scanning_nested(self):
        placeholders = get_placeholders('placeholder_tests/test_six.html')
        self.assertEqual(sorted(placeholders), sorted([u'new_one', u'new_two', u'new_three']))

#    def test_placeholder_scanning_duplicate(self):
#        placeholders = self.assertWarns(DuplicatePlaceholderWarning, "Duplicate placeholder found: `one`", get_placeholders, 'placeholder_tests/test_seven.html')
#        self.assertEqual(sorted(placeholders), sorted([u'one']))

    def test_placeholder_scanning_extend_outside_block(self):
        placeholders = get_placeholders('placeholder_tests/outside.html')
        self.assertEqual(sorted(placeholders), sorted([u'new_one', u'two', u'base_outside']))

    def test_placeholder_scanning_extend_outside_block_nested(self):
        placeholders = get_placeholders('placeholder_tests/outside_nested.html')
        self.assertEqual(sorted(placeholders), sorted([u'new_one', u'two', u'base_outside']))

    def test_placeholder_scanning_nested_super(self):
        placeholders = get_placeholders('placeholder_tests/nested_super_level1.html')
        self.assertEqual(sorted(placeholders), sorted([u'level1', u'level2', u'level3', u'level4']))

    def test_ignore_variable_extends(self):
        placeholders = get_placeholders('placeholder_tests/variable_extends.html')
        self.assertEqual(sorted(placeholders), [])

    def test_tag_placeholder_exception(self):
        exp = TemplateSyntaxError('placeholder tag requires 2 arguments')
        with self.assertRaises(TemplateSyntaxError) as tsx:
            get_placeholders('placeholder_tests/tag_exception.html')
        self.assertEqual(str(tsx.exception), str(exp))


template_analyzer/tests/__init__.py
from .test_placeholder import PlaceholderTestCase  # Expose for Django < 1.6


template_analyzer/tests/app_loader.py
"""
Based on:
* https://github.com/django-admin-tools/django-admin-tools/blob/master/admin_tools/template_loaders.py
* https://bitbucket.org/tzulberti/django-apptemplates/
* http://djangosnippets.org/snippets/1376/

Django template loader that allows you to load a template from a
specific application. This allows you to both extend and override
a template at the same time. The default Django loaders require you
to copy the entire template you want to override, even if you only
want to override one small block.

Template usage example::
    {% extends "admin:admin/base.html" %}
"""
from importlib import import_module
from os.path import abspath, dirname, join

import django
from django.apps import apps
from django.template import Origin
from django.template.loaders.filesystem import Loader as FilesystemLoader

_cache = {}


def get_app_template_dir(app_name):
    """Get the template directory for an application

    Uses apps interface available in django 1.7+

    Returns a full path, or None if the app was not found.
    """
    if app_name in _cache:
        return _cache[app_name]
    template_dir = None

    for app in apps.get_app_configs():
        if app.label == app_name:
            template_dir = join(app.path, "templates")
            break

    _cache[app_name] = template_dir
    return template_dir


class Loader(FilesystemLoader):
    is_usable = True

    def get_template_sources(self, template_name, template_dirs=None):
        """
        Returns the absolute paths to "template_name" in the specified app.
        If the name does not contain an app name (no colon), an empty list
        is returned.
        The parent FilesystemLoader.load_template_source() will take care
        of the actual loading for us.
        """
        if ":" not in template_name:
            return []
        app_name, template_name = template_name.split(":", 1)
        template_dir = get_app_template_dir(app_name)
        if template_dir:
            origin = Origin(
                name=join(template_dir, template_name),
                template_name=template_name,
                loader=self,
            )
            return [origin]
        return []


runtests.py
#!/usr/bin/env python
import sys

import django
from django.conf import settings
from django.core.management import execute_from_command_line

if not settings.configured:
    settings.configure(
        DEBUG=True,
        DATABASES={"default": {"ENGINE": "django.db.backends.sqlite3", "NAME": ":memory:"}},
        INSTALLED_APPS=("template_analyzer",),
        MIDDLEWARE=(),
        TEMPLATES=[
            {
                "BACKEND": "django.template.backends.django.DjangoTemplates",
                "DIRS": (),
                "OPTIONS": {
                    "loaders": (
                        "django.template.loaders.filesystem.Loader",
                        "django.template.loaders.app_directories.Loader",
                        # added dynamically: 'template_analyzer.tests.app_loader.Loader',
                    ),
                },
            },
        ],
    )


def runtests():
    argv = sys.argv[:1] + ["test", "template_analyzer", "--traceback"] + sys.argv[1:]
    execute_from_command_line(argv)


if __name__ == "__main__":
    runtests()


cms/utils/plugins.py
# -*- coding: utf-8 -*-
from cms.exceptions import DuplicatePlaceholderWarning
from cms.models import Page
from cms.templatetags.cms_tags import Placeholder
from cms.utils.placeholder import validate_placeholder_name
from django.contrib.sites.models import Site
from django.shortcuts import get_object_or_404
from django.template import (NodeList, TextNode, VariableNode,
    TemplateSyntaxError)
from django.template.loader import get_template
from django.template.loader_tags import (ConstantIncludeNode, ExtendsNode,
    BlockNode)
import warnings

def get_page_from_plugin_or_404(cms_plugin):
    return get_object_or_404(Page, placeholders=cms_plugin.placeholder)

def _extend_blocks(extend_node, blocks):
    """
    Extends the dictionary `blocks` with *new* blocks in the parent node (recursive)
    """
    # we don't support variable extensions
    if extend_node.parent_name_expr:
        return
    parent = extend_node.get_parent(None)
    # Search for new blocks
    for node in parent.nodelist.get_nodes_by_type(BlockNode):
        if not node.name in blocks:
            blocks[node.name] = node
        else:
            # set this node as the super node (for {{ block.super }})
            block = blocks[node.name]
            seen_supers = []
            while hasattr(block.super, 'nodelist') and block.super not in seen_supers:
                seen_supers.append(block.super)
                block = block.super
            block.super = node
    # search for further ExtendsNodes
    for node in parent.nodelist.get_nodes_by_type(ExtendsNode):
        _extend_blocks(node, blocks)
        break

def _find_topmost_template(extend_node):
    parent_template = extend_node.get_parent({})
    for node in parent_template.nodelist.get_nodes_by_type(ExtendsNode):
        # Their can only be one extend block in a template, otherwise django raises an exception
        return _find_topmost_template(node)
    # No ExtendsNode
    return extend_node.get_parent({})

def _extend_nodelist(extend_node):
    """
    Returns a list of placeholders found in the parent template(s) of this
    ExtendsNode
    """
    # we don't support variable extensions
    if extend_node.parent_name_expr:
        return []
    blocks = extend_node.blocks
    _extend_blocks(extend_node, blocks)
    placeholders = []

    for block in blocks.values():
        placeholders += _scan_placeholders(block.nodelist, block, blocks.keys())

    # Scan topmost template for placeholder outside of blocks
    parent_template = _find_topmost_template(extend_node)
    placeholders += _scan_placeholders(parent_template.nodelist, None, blocks.keys())
    return placeholders

def _scan_placeholders(nodelist, current_block=None, ignore_blocks=[]):
    placeholders = []

    for node in nodelist:
        # check if this is a placeholder first
        if isinstance(node, Placeholder):
            placeholders.append(node.get_name())
        # if it's a Constant Include Node ({% include "template_name.html" %})
        # scan the child template
        elif isinstance(node, ConstantIncludeNode):
            # if there's an error in the to-be-included template, node.template becomes None
            if node.template:
                placeholders += _scan_placeholders(node.template.nodelist, current_block)
        # handle {% extends ... %} tags
        elif isinstance(node, ExtendsNode):
            placeholders += _extend_nodelist(node)
        # in block nodes we have to scan for super blocks
        elif isinstance(node, VariableNode) and current_block:
            if node.filter_expression.token == 'block.super':
                if not hasattr(current_block.super, 'nodelist'):
                    raise TemplateSyntaxError("Cannot render block.super for blocks without a parent.")
                placeholders += _scan_placeholders(current_block.super.nodelist, current_block.super)
        # ignore nested blocks which are already handled
        elif isinstance(node, BlockNode) and node.name in ignore_blocks:
            continue
        # if the node has the newly introduced 'child_nodelists' attribute, scan
        # those attributes for nodelists and recurse them
        elif hasattr(node, 'child_nodelists'):
            for nodelist_name in node.child_nodelists:
                if hasattr(node, nodelist_name):
                    subnodelist = getattr(node, nodelist_name)
                    if isinstance(subnodelist, NodeList):
                        if isinstance(node, BlockNode):
                            current_block = node
                        placeholders += _scan_placeholders(subnodelist, current_block)
        # else just scan the node for nodelist instance attributes
        else:
            for attr in dir(node):
                obj = getattr(node, attr)
                if isinstance(obj, NodeList):
                    if isinstance(node, BlockNode):
                        current_block = node
                    placeholders += _scan_placeholders(obj, current_block)
    return placeholders

def get_placeholders(template):
    compiled_template = get_template(template)
    placeholders = _scan_placeholders(compiled_template.nodelist)
    clean_placeholders = []
    for placeholder in placeholders:
        if placeholder in clean_placeholders:
            warnings.warn("Duplicate placeholder found: `%s`" % placeholder, DuplicatePlaceholderWarning)
        else:
            validate_placeholder_name(placeholder)
            clean_placeholders.append(placeholder)
    return clean_placeholders

SITE_VAR = "site__exact"

def current_site(request):
    if SITE_VAR in request.REQUEST:
        return Site.objects.get(pk=request.REQUEST[SITE_VAR])
    else:
        site_pk = request.session.get('cms_admin_site', None)
        if site_pk:
            try:
                return Site.objects.get(pk=site_pk)
            except Site.DoesNotExist:
                return None
        else:
            return Site.objects.get_current()


template_analyzer/templatetags/template_analyzer_test_tags.py
from django.template import Library, Node, TemplateSyntaxError, Variable

register = Library()


class Placeholder(Node):
    """
    Simple placeholder node.
    """

    @classmethod
    def parse(cls, parser, token):
        tokens = token.contents.split()
        if len(tokens) == 2:
            return cls(Variable(tokens[1]))
        else:
            raise TemplateSyntaxError("{} tag requires 2 arguments".format(tokens[0]))

    def __init__(self, name_var):
        self.name_var = name_var

    def get_name(self):
        return self.name_var.literal

    def render(self, context):
        return "[placeholder: {}]".format(self.name_var.literal)


@register.tag
def placeholder(parser, token):
    """
    A dummy placeholder template tag.
    """
    return Placeholder.parse(parser, token)


template_analyzer/djangoanalyzer.py
# Ensure that loader is imported before loader_tags, or a circular import may happen
# when this file is loaded from an `__init__.py` in an application root.
# The __init__.py files in applications are loaded very early due to the scan of of translation.activate()
import django
import django.template.loader  # noqa

# Normal imports
from django.template import Context, NodeList, Template, TemplateSyntaxError
from django.template.backends.django import Template as TemplateAdapter
from django.template.base import VariableNode
from django.template.loader import get_template
from django.template.loader_tags import BlockNode, ExtendsNode, IncludeNode


def _is_variable_extends(extend_node):
    """
    Check whether an ``{% extends variable %}`` is used in the template.

    :type extend_node: ExtendsNode
    """
    # The FilterExpression.var can be either a string, or Variable object.
    return not isinstance(extend_node.parent_name.var, str)


def _extend_blocks(extend_node, blocks, context):
    """
    Extends the dictionary `blocks` with *new* blocks in the parent node (recursive)

    :param extend_node: The ``{% extends .. %}`` node object.
    :type extend_node: ExtendsNode
    :param blocks: dict of all block names found in the template.
    :type blocks: dict
    """
    try:
        # This needs a fresh parent context, or it will detection recursion in Django 1.9+,
        # and thus skip the base template, which is already loaded.
        parent = extend_node.get_parent(_get_extend_context(context))
    except TemplateSyntaxError:
        if _is_variable_extends(extend_node):
            # we don't support variable extensions unless they have a default.
            return
        else:
            raise

    # Search for new blocks
    for parent_block in parent.nodelist.get_nodes_by_type(BlockNode):
        if not parent_block.name in blocks:
            blocks[parent_block.name] = parent_block
        else:
            # set this node as the super node (for {{ block.super }})
            block = blocks[parent_block.name]
            seen_supers = []
            while hasattr(block.parent, "nodelist") and block.parent not in seen_supers:
                seen_supers.append(block.parent)
                block = block.parent
            block.parent = parent_block

    # search for further ExtendsNodes in the extended template
    # There is only one extend block in a template (Django checks for this).
    parent_extends = parent.nodelist.get_nodes_by_type(ExtendsNode)
    if parent_extends:
        _extend_blocks(parent_extends[0], blocks, context)


def _find_topmost_template(extend_node, context):
    try:
        parent_template = extend_node.get_parent(context)
    except TemplateSyntaxError:
        # we don't support variable extensions
        if _is_variable_extends(extend_node):
            return
        else:
            raise

    # There is only one extend block in a template (Django checks for this).
    parent_extends = parent_template.nodelist.get_nodes_by_type(ExtendsNode)
    if parent_extends:
        return _find_topmost_template(parent_extends[0], context)
    else:
        # No ExtendsNode
        return parent_template


def _extend_nodelist(extends_node, context, instance_types):
    """
    Returns a list of results found in the parent template(s)
    :type extends_node: ExtendsNode
    """
    results = []

    # Find all blocks in the complete inheritance chain
    blocks = extends_node.blocks.copy()  # dict with all blocks in the current template
    _extend_blocks(extends_node, blocks, context)

    # Dive into all blocks of the page one by one
    all_block_names = list(blocks.keys())
    for block in list(blocks.values()):
        results += _scan_nodes(
            block.nodelist, context, instance_types, block, ignore_blocks=all_block_names
        )

    # Scan topmost template for nodes that exist outside of blocks
    parent_template = _find_topmost_template(extends_node, context)
    if not parent_template:
        return []
    else:
        results += _scan_nodes(
            parent_template.nodelist, context, instance_types, ignore_blocks=all_block_names
        )
        return results


def _scan_nodes(nodelist, context, instance_types, current_block=None, ignore_blocks=None):
    """
    Loop through all nodes of a single scope level.

    :type nodelist: django.template.base.NodeList
    :type current_block: BlockNode
    :param instance_types: The instance to look for
    """
    results = []
    for node in nodelist:
        # first check if this is the object instance to look for.
        if isinstance(node, instance_types):
            results.append(node)
            # if it's a Constant Include Node ({% include "template_name.html" %})
        # scan the child template
        elif isinstance(node, IncludeNode):
            # if there's an error in the to-be-included template, node.template becomes None
            if node.template:
                # This is required for Django 1.7 but works on older version too
                # Check if it quacks like a template object, if not
                # presume is a template path and get the object out of it
                if not callable(getattr(node.template, "render", None)):
                    template = get_template(node.template.var)
                else:
                    template = node.template

                if isinstance(template, TemplateAdapter):
                    # Django 1.8+: received a new object, take original template
                    template = template.template

                results += _scan_nodes(template.nodelist, context, instance_types, current_block)
        # handle {% extends ... %} tags
        elif isinstance(node, ExtendsNode):
            results += _extend_nodelist(node, context, instance_types)
        # in block nodes we have to scan for super blocks
        elif isinstance(node, VariableNode) and current_block:
            if node.filter_expression.token == "block.super":
                # Found a {{ block.super }} line
                if not hasattr(current_block.parent, "nodelist"):
                    raise TemplateSyntaxError(
                        "Cannot read {{{{ block.super }}}} for {{% block {0} %}}, "
                        "the parent template doesn't have this block.".format(current_block.name)
                    )
                results += _scan_nodes(
                    current_block.parent.nodelist, context, instance_types, current_block.parent
                )
        # ignore nested blocks which are already handled
        elif isinstance(node, BlockNode) and ignore_blocks and node.name in ignore_blocks:
            continue
        # if the node has the newly introduced 'child_nodelists' attribute, scan
        # those attributes for nodelists and recurse them
        elif hasattr(node, "child_nodelists"):
            for nodelist_name in node.child_nodelists:
                if hasattr(node, nodelist_name):
                    subnodelist = getattr(node, nodelist_name)
                    if isinstance(subnodelist, NodeList):
                        if isinstance(node, BlockNode):
                            current_block = node
                        results += _scan_nodes(subnodelist, context, instance_types, current_block)
        # else just scan the node for nodelist instance attributes
        else:
            for attr in dir(node):
                obj = getattr(node, attr)
                if isinstance(obj, NodeList):
                    if isinstance(node, BlockNode):
                        current_block = node
                    results += _scan_nodes(obj, context, instance_types, current_block)
    return results


def _get_main_context(nodelist):
    # The context is empty, but needs to be provided to handle the {% extends %} node.
    context = Context({})

    if isinstance(nodelist, TemplateAdapter):
        # The top-level context.
        context.template = Template("", engine=nodelist.template.engine)
    else:
        # Just in case a different nodelist is provided.
        # Using the default template now.
        context.template = Template("")
    return context


def _get_extend_context(parent_context):
    # For extends nodes, a fresh template instance is constructed.
    # The loader cache of the original `nodelist` is skipped.
    context = Context({})
    context.template = Template("", engine=parent_context.template.engine)
    return context


def get_node_instances(nodelist, instances):
    """
    Find the nodes of a given instance.

    In contract to the standard ``template.nodelist.get_nodes_by_type()`` method,
    this also looks into ``{% extends %}`` and ``{% include .. %}`` nodes
    to find all possible nodes of the given type.

    :param instances: A class Type, or tuple of types to find.
    :param nodelist:  The Template object, or nodelist to scan.
    :returns: A list of Node objects which inherit from the list of given `instances` to find.
    :rtype: list
    """
    context = _get_main_context(nodelist)

    # The Django loader returns an adapter class;
    # it wraps the original Template in a new object to be API compatible
    if isinstance(nodelist, TemplateAdapter):
        nodelist = nodelist.template

    if isinstance(nodelist, Template):
        # As of Django 4.1, template Node objects no longer allow iteration,
        # which breaks Template.__iter__ too. Instead, directly walk over the nodelist.
        nodelist = nodelist.nodelist

    return _scan_nodes(nodelist, context, instances)
